File: react-native-nitro-player/ios/queue/QueueManager.swift
//
//  QueueManager.swift
//  NitroPlayer
//
//  Created on 24/11/25.
//

import Foundation

/**
 * Wrapper class to store listeners with unique IDs for removal
 */
private class ListenerWrapper {
    let id: UUID
    let listener: ([TrackItem], QueueOperation?) -> Void
    
    init(id: UUID = UUID(), listener: @escaping ([TrackItem], QueueOperation?) -> Void) {
        self.id = id
        self.listener = listener
    }
}

/**
 * QueueManager is a singleton that manages the queue state across the app session.
 * It provides thread-safe access to the queue and notifies listeners of queue changes.
 */
class QueueManager {
    private let queue = Queue()
    private var listeners: [ListenerWrapper] = []
    private let listenersQueue = DispatchQueue(label: "com.margelo.nitro.nitroplayer.listeners")
    
    static let shared = QueueManager()
    
    private init() {}
    
    /**
     * Get the current queue
     */
    func getQueue() -> Queue {
        return queue
    }
    
    /**
     * Get all tracks in the queue
     */
    func getTracks() -> [TrackItem] {
        return queue.getTracks()
    }
    
    /**
     * Get tracks as an array
     */
    func getTracksArray() -> [TrackItem] {
        return queue.getTracksArray()
    }
    
    /**
     * Load multiple tracks into the queue (replaces existing queue)
     */
    func loadQueue(_ tracks: [TrackItem]) {
        queue.loadTracks(tracks)
        notifyListeners(.add)
    }
    
    /**
     * Load a single track at a specific index
     */
    func loadSingleTrack(_ track: TrackItem, index: Double?) {
        let insertIndex = index.map { Int($0) }
        if let insertIndex = insertIndex, insertIndex >= 0 {
            queue.addTrackAtIndex(track, index: insertIndex)
        } else {
            queue.addTrack(track)
        }
        notifyListeners(.add)
    }
    
    /**
     * Delete a track by ID
     */
    func deleteTrack(id: String) {
        let removed = queue.removeTrack(id: id)
        if removed {
            notifyListeners(.remove)
        }
    }
    
    /**
     * Clear all tracks from the queue
     */
    func clearQueue() {
        queue.clear()
        notifyListeners(.clear)
    }
    
    /**
     * Add a listener for queue changes
     * @param listener Callback that receives (queue, operation)
     * @return A function to remove the listener
     */
    func addQueueChangeListener(_ listener: @escaping (_ queue: [TrackItem], _ operation: QueueOperation?) -> Void) -> () -> Void {
        let wrapper = ListenerWrapper(listener: listener)
        listenersQueue.sync {
            listeners.append(wrapper)
        }
        
        return { [weak self, wrapper] in
            self?.listenersQueue.sync {
                self?.listeners.removeAll { $0.id == wrapper.id }
            }
        }
    }
    
    /**
     * Remove a queue change listener
     */
    func removeQueueChangeListener(_ listener: @escaping (_ queue: [TrackItem], _ operation: QueueOperation?) -> Void) {
        listenersQueue.sync {
            // This is kept for API compatibility, but removal should use the returned function
            // from addQueueChangeListener
        }
    }
    
    /**
     * Notify all listeners of queue changes
     */
    private func notifyListeners(_ operation: QueueOperation?) {
        let currentTracks = queue.getTracks()
        listenersQueue.sync {
            let listenersCopy = listeners
            for wrapper in listenersCopy {
                do {
                    wrapper.listener(currentTracks, operation)
                } catch {
                    // Log error but don't break other listeners
                    print("Error in queue change listener: \(error)")
                }
            }
        }
    }
    
    /**
     * Get queue size
     */
    func getQueueSize() -> Int {
        return queue.size()
    }
    
    /**
     * Check if queue is empty
     */
    func isQueueEmpty() -> Bool {
        return queue.isEmpty()
    }
    
    /**
     * Get a track by index
     */
    func getTrack(index: Int) -> TrackItem? {
        return queue.getTrack(index: index)
    }
    
    /**
     * Get a track by ID
     */
    func getTrackById(id: String) -> TrackItem? {
        return queue.getTrackById(id: id)
    }
    
    /**
     * Get the index of a track by ID
     */
    func getTrackIndex(id: String) -> Int {
        return queue.getTrackIndex(id: id)
    }
}

File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/queue/QueueManager.kt
package com.margelo.nitro.nitroplayer.queue

import com.margelo.nitro.nitroplayer.QueueOperation
import com.margelo.nitro.nitroplayer.TrackItem
import java.util.concurrent.CopyOnWriteArrayList

/**
 * QueueManager is a singleton that manages the queue state across the app session.
 * It provides thread-safe access to the queue and notifies listeners of queue changes.
 */
class QueueManager private constructor() {
    
    private val queue = Queue()
    private val listeners = CopyOnWriteArrayList<(List<TrackItem>, QueueOperation?) -> Unit>()
    
    companion object {
        @Volatile
        private var INSTANCE: QueueManager? = null
        
        /**
         * Get the singleton instance of QueueManager
         */
        @JvmStatic
        fun getInstance(): QueueManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: QueueManager().also { INSTANCE = it }
            }
        }
    }
    
    /**
     * Get the current queue
     */
    fun getQueue(): Queue {
        return queue
    }
    
    /**
     * Get all tracks in the queue
     */
    fun getTracks(): List<TrackItem> {
        return queue.getTracks()
    }
    
    /**
     * Get tracks as an array
     */
    fun getTracksArray(): Array<TrackItem> {
        return queue.getTracksArray()
    }
    
    /**
     * Load multiple tracks into the queue (replaces existing queue)
     */
    fun loadQueue(tracks: Array<TrackItem>) {
        queue.loadTracks(tracks.toList())
        notifyListeners(QueueOperation.ADD)
        
        // Update MediaBrowserService cache immediately
        try {
            val service = com.margelo.nitro.nitroplayer.media.NitroPlayerMediaBrowserService.getInstance()
            service?.updateQueue(queue.getTracks())
            println("📋 QueueManager: Updated MediaBrowserService cache with ${queue.getTracks().size} tracks")
        } catch (e: Exception) {
            println("⚠️ QueueManager: Error updating MediaBrowserService cache: ${e.message}")
        }
    }
    
    /**
     * Load a single track at a specific index
     */
    fun loadSingleTrack(track: TrackItem, index: Double?) {
        val insertIndex = index?.toInt()
        if (insertIndex != null && insertIndex >= 0) {
            queue.addTrackAtIndex(track, insertIndex)
        } else {
            queue.addTrack(track)
        }
        notifyListeners(QueueOperation.ADD)
        
        // Update MediaBrowserService cache
        try {
            val service = com.margelo.nitro.nitroplayer.media.NitroPlayerMediaBrowserService.getInstance()
            service?.updateQueue(queue.getTracks())
        } catch (e: Exception) {
            println("⚠️ QueueManager: Error updating MediaBrowserService cache: ${e.message}")
        }
    }
    
    /**
     * Delete a track by ID
     */
    fun deleteTrack(id: String) {
        val removed = queue.removeTrack(id)
        if (removed) {
            notifyListeners(QueueOperation.REMOVE)
        }
    }
    
    /**
     * Clear all tracks from the queue
     */
    fun clearQueue() {
        queue.clear()
        notifyListeners(QueueOperation.CLEAR)
    }
    
    /**
     * Add a listener for queue changes
     * @param listener Callback that receives (queue, operation)
     * @return A function to remove the listener
     */
    fun addQueueChangeListener(listener: (List<TrackItem>, QueueOperation?) -> Unit): () -> Unit {
        listeners.add(listener)
        return { listeners.remove(listener) }
    }
    
    /**
     * Remove a queue change listener
     */
    fun removeQueueChangeListener(listener: (List<TrackItem>, QueueOperation?) -> Unit) {
        listeners.remove(listener)
    }
    
    /**
     * Notify all listeners of queue changes
     */
    private fun notifyListeners(operation: QueueOperation?) {
        val currentTracks = queue.getTracks()
        listeners.forEach { listener ->
            try {
                listener(currentTracks, operation)
            } catch (e: Exception) {
                // Log error but don't break other listeners
                e.printStackTrace()
            }
        }
    }
    
    /**
     * Get queue size
     */
    fun getQueueSize(): Int {
        return queue.size()
    }
    
    /**
     * Check if queue is empty
     */
    fun isQueueEmpty(): Boolean {
        return queue.isEmpty()
    }
    
    /**
     * Get a track by index
     */
    fun getTrack(index: Int): TrackItem? {
        return queue.getTrack(index)
    }
    
    /**
     * Get a track by ID
     */
    fun getTrackById(id: String): TrackItem? {
        return queue.getTrackById(id)
    }
    
    /**
     * Get the index of a track by ID
     */
    fun getTrackIndex(id: String): Int {
        return queue.getTrackIndex(id)
    }
}

